### PR TITLE
ext.bridge improvements

### PR DESCRIPTION
**Pull request by Middledot**
## Summary
Makes a bunch of changes to ext.bridge
- Implements bridge command groups
    - `bridge.map_to` decorator to use main command as subcommand
- Attachment option
- `is_app` property added to better differentiate context types

## Information
- [x] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed).
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting, examples, ...).

## Checklist
- [x] I have searched the open pull requests for duplicates.
- [x] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why.

### CODE REVIEW

File: discord/commands/core.py [hunk 1]
```diff
@@ -1010,10 +1010,10 @@ def __init__(
...
         **kwargs,
     ) -> None:
-        validate_chat_input_name(name)
-        validate_chat_input_description(description)
-        self.name = str(name)
-        self.description = description
+        self.name = name
+        self.description = description or "No description provided"
+        validate_chat_input_name(self.name)
+        validate_chat_input_description(self.description)
```
Comment (Middledot) [inline discord/commands/core.py:1016]: If description was set as null, a validation error would be raised

File: discord/ext/bridge/core.py [hunk 2]
```diff
@@ -22,98 +22,74 @@
...
-    kwargs: Optional[Dict[:class:`str`, Any]]
-        Keyword arguments that are directly passed to the respective command constructors.
-    """
+def filter_params(params, **kwargs):
+    for param, nparam in kwargs.items():
+        if param in params:
+            if nparam is None:
+                params.pop(param)
+            else:
+                kwargs[nparam] = params.pop(param)
```
Comment (Middledot) [inline discord/ext/bridge/core.py:63]: Could be moved to utils?

Comment (Lulalaby) [reply discord/ext/bridge/core.py:64]: Yeah

File: discord/ext/bridge/core.py [hunk 3]
```diff
@@ -192,31 +160,156 @@ def after_invoke(self, coro):
...
+    kwargs: Optional[Dict[:class:`str`, Any]]
+        Keyword arguments that are directly passed to the respective command constructors. (:class:`.SlashCommand` and :class:`.ext.commands.Command`)
+    """
+    def __init__(self, callback, **kwargs):
+        self.slash_variant = kwargs.pop("slash_variant", None) or BridgeSlashCommand(callback, **kwargs)
+        self.ext_variant = kwargs.pop("ext_variant", None) or BridgeExtCommand(callback, **kwargs)
+
+    def add_to(self, bot: ExtBot) -> None:
+        bot.add_application_command(self.slash_variant)
+        bot.add_command(self.ext_variant)
```
[the remark this answers is not in the document]
Comment (Lulalaby) [reply discord/ext/bridge/core.py:232]: We should still document it I think

[the remark this answers is not in the document]
Comment (Dorukyum) [reply discord/ext/bridge/core.py:232]: I agree, it's better to keep the doc

File: discord/ext/bridge/core.py (hunk 2 above)
Comment (Dorukyum) [reply discord/ext/bridge/core.py:64]: If you could add a docstring explaining what this is and why it's useful :)

**Review by Lulalaby — APPROVED** [review]

**Comment by Lulalaby** [comment]
@Dorukyum

File: discord/ext/bridge/context.py [hunk 4]
```diff
@@ -140,6 +140,10 @@ async def _defer(self, *args, **kwargs) -> None:
     async def _edit(self, *args, **kwargs) -> InteractionMessage:
         return await self._get_super("edit")(*args, **kwargs)
 
+    @property
+    def is_app(self) -> bool:
+        return True
```
Comment (Dorukyum) [inline discord/ext/bridge/context.py:145]: Since this value never changes, it can be replaced with a classvar

File: discord/ext/bridge/context.py [hunk 5]
```diff
@@ -184,3 +188,7 @@ async def delete(
         """
         if self._original_response_message:
             await self._original_response_message.delete(delay=delay, reason=reason)
+
+    @property
+    def is_app(self) -> bool:
+        return False
```
Comment (Dorukyum) [inline discord/ext/bridge/context.py:194]: Since this value never changes, it can be replaced with a classvar

File: discord/ext/bridge/core.py [hunk 6]
```diff
@@ -192,31 +189,163 @@ def after_invoke(self, coro):
...
+    mapped: Optional[:class:`.SlashCommand`]
+        If :func:`map_to` is used, the mapped slash command.
     """
+    def __init__(self, callback, *args, **kwargs):
+        self.ext_variant = Group(callback, *args, **kwargs)
+        self.slash_variant = SlashCommandGroup(self.ext_variant.name, *args, **kwargs)
+        self.subcommands: List[BridgeCommand] = []
+
+        self.mapped = None
+        if (map_to := getattr(callback, "__custom_map_to__", None)):
```
Comment (Dorukyum) [inline discord/ext/bridge/core.py:260]:
```suggestion
        if map_to := getattr(callback, "__custom_map_to__", None):
```

File: discord/utils.py [hunk 7]
```diff
@@ -1164,3 +1164,35 @@ def check(item: Any) -> bool:
         return iter(itertools.islice(gen, 25))
 
     return autocomplete_callback
+
+
+def filter_params(params, **kwargs):
+    """Mainly used in ext.bridge, filter out and replace certain keyword parameters
```
Comment (Dorukyum) [inline discord/utils.py:1170]:
```suggestion
    """A helper function to filter out and replace certain keyword parameters.
```

**Review by Dorukyum — DISMISSED** [review]

**Comment by Dorukyum** [comment]
I'm facing two issues. The code I used for testing:
```py
bot = bridge.Bot(command_prefix="t.", intents=Intents.all())

@bot.bridge_group(description="Test command")
async def hello(ctx: bridge.BridgeContext, attachment: Attachment):
    await ctx.respond(attachment.filename)
```

1. The slash command doesn't have any options
2. I get the following exception even though I have attached a file:
```
Ignoring exception in command hello:
Traceback (most recent call last):
  File "/home/dorukyum/Desktop/Code/pycord/discord/ext/commands/bot.py", line 344, in invoke
    await ctx.command.invoke(ctx)
  File "/home/dorukyum/Desktop/Code/pycord/discord/ext/commands/core.py", line 1532, in invoke
    await self.prepare(ctx)
  File "/home/dorukyum/Desktop/Code/pycord/discord/ext/commands/core.py", line 873, in prepare
    await self._parse_arguments(ctx)
  File "/home/dorukyum/Desktop/Code/pycord/discord/ext/commands/core.py", line 779, in _parse_arguments
    transformed = await self.transform(ctx, param)
  File "/home/dorukyum/Desktop/Code/pycord/discord/ext/commands/core.py", line 612, in transform
    raise MissingRequiredArgument(param)
discord.ext.commands.errors.MissingRequiredArgument: attachment is a required argument that is missing.
```

**Comment by Middledot** [comment]
Probably fixed now. I just made it so that slash groups without subcommands are not registered since the library in general doesn't support slash groups with callbacks.

File: discord/interactions.py [hunk 8]
```diff
@@ -1,5 +1,3 @@
-# -*- coding: utf-8 -*-
```
Comment (EmreTech) [inline discord/interactions.py:1]: I would recommend making changes that fit the context of this PR

File: discord/utils.py [hunk 9]
```diff
@@ -1164,3 +1164,35 @@ def check(item: Any) -> bool:
...
+    .. code-block:: python3
+
+        >>> params = {"param1": 12, "param2": 13}
+        >>> filter_params(params, param1="param3", param2=None)
+        {'param3': 12}
+        # values of 'param1' is moved to 'param3'
+        # and values of 'param2' are completely removed.
+
+    """
+    for param, nparam in kwargs.items():
```
Comment (EmreTech) [inline discord/utils.py:1191]: `nparam` is a confusing variable name. `param_value` sounds better

File: discord/ext/bridge/core.py [hunk 10]
```diff
@@ -192,31 +201,170 @@ def after_invoke(self, coro):
...
+    kwargs: Optional[Dict[:class:`str`, Any]]
+        Keyword arguments that are directly passed to the respective command constructors. (:class:`.SlashCommandGroup` and :class:`.ext.commands.Group`)
+    """
+    def decorator(callback):
+        return BridgeCommandGroup(callback, **kwargs)
+
+    return decorator
+
+
+def map_to(name, description = None):
```
Comment (EmreTech) [inline discord/ext/bridge/core.py:329]: I don't get the point of this function. Why is it in `ext.bridge` if it's only supposed to be used on Slash Commands?

File: discord/ext/bridge/core.py [hunk 11]
```diff
@@ -22,98 +22,115 @@
...
-    def get_ext_command(self):
-        """A method to get the ext.commands version of this command.
 
-        Returns
-        -------
-        :class:`BridgeExtCommand`
-            The respective traditional (prefix-based) version of the command.
-        """
-        return self.ext_command
+class BaseBridgeCommand:
```
Comment (EmreTech) [inline discord/ext/bridge/core.py:93]: Is this really necessary? This base class is only inherited by `BridgeCommand`

File: discord/ext/bridge/context.py [hunk 12]
```diff
@@ -50,14 +50,15 @@ class BridgeContext(ABC):
...
-            elif isinstance(ctx, BridgeApplicationContext):
+            if ctx.is_app:
                 command_type = "Application command"
+            else:
+                command_type = "Traditional (prefix-based) command"
             await ctx.send(f"This command was invoked with a(n) {command_type}.")
 
     .. versionadded:: 2.0
     """
+    is_app: bool
```
Comment (EmreTech) [inline discord/ext/bridge/context.py:61]: Alternative idea for this: make `is_app` a property, and return the value of `isinstance(self, BridgeApplicationContext)`

File: discord/commands/core.py [hunk 13]
```diff
@@ -1016,10 +1016,10 @@ def __init__(
...
```
Comment (EmreTech) [inline discord/commands/core.py:1022]: How are these changes related to the topic of this PR?

File: discord/ext/bridge/core.py (hunk 10 above)
Comment (Middledot) [reply discord/ext/bridge/core.py:329]: Here it's supposed to map the primary command used by `Group` to a subcommand for slash so that it isn't wasted. Regular slash command groups don't have callbacks so if you mean that it could be used for regular groups that could be done in another pr.

File: discord/commands/core.py (hunk 13 above)
Comment (Middledot) [reply discord/commands/core.py:1022]: I found the bug while working on groups. I'll move it to another pr

File: discord/utils.py (hunk 9 above)
Comment (Middledot) [reply discord/utils.py:1191]: `nparam` was supposed to be the new param so I'll name it that

**Review by EmreTech — DISMISSED** [review]
LGTM, haven't tested

**Review by EmreTech — APPROVED** [review]
Pain

**Comment by Lulalaby** [comment]
> Pain

Hahahahaha

I won't approve this tho

**Comment by Dorukyum** [comment]
Could we add an example?

**Comment by Dorukyum** [comment]
I get the following error on startup now:
```
Traceback (most recent call last):
  File "test.py", line 9, in <module>
    async def hello(ctx: bridge.BridgeContext):
  File "discord/ext/bridge/bot.py", line 88, in decorator
    result = bridge_group(**kwargs)(func)
  File "discord/ext/bridge/core.py", line 318, in decorator
    return BridgeCommandGroup(callback, **kwargs)
  File "discord/ext/bridge/core.py", line 262, in __init__
    self.slash_variant: BridgeSlashGroup = BridgeSlashGroup(self.ext_variant.name, *args, **kwargs)
  File "discord/ext/bridge/core.py", line 64, in __init__
    super().__init__(func, **kwargs)
  File "discord/commands/core.py", line 625, in __init__
    raise TypeError("Callback must be a coroutine.")
TypeError: Callback must be a coroutine.
```
The code I used:
```py
bot = bridge.Bot(command_prefix="t.", intents=Intents.all())

@bot.bridge_group(description="Test command")
async def hello(ctx: bridge.BridgeContext):
    ...

@hello.command()
async def there(ctx: bridge.BridgeContext, attachment: Attachment):
    await ctx.respond(attachment.filename)
```

File: discord/commands/core.py [hunk 14]
```diff
@@ -1065,9 +1065,9 @@ def to_dict(self) -> Dict:
 
         return as_dict
 
-    def command(self, **kwargs) -> Callable[[Callable], SlashCommand]:
-        def wrap(func) -> SlashCommand:
-            command = SlashCommand(func, parent=self, **kwargs)
+    def command(self, cls: Type[T] = SlashCommand, **kwargs) -> Callable[[Callable], SlashCommand]:
+        def wrap(func) -> T:
+            command = cls(func, parent=self, **kwargs)
```
Comment (Middledot) [inline discord/commands/core.py:1070]: Not too sure about the typing here

**Review by Middledot — COMMENTED** [review]
Should be fixed and an example was added.

File: discord/ext/bridge/core.py [hunk 15]
```diff
@@ -226,9 +383,14 @@ async def convert(self, ctx, argument):
...
             return await UserConverter().convert(ctx, argument)
 
-
-def attachment_callback(*args):  # pylint: disable=unused-argument
-    raise ValueError("Attachments are not supported for bridge commands.")
+class AttachmentConverter(Converter):
+    async def convert(self, ctx: Context, arg: str):
+        try:
+            attach = ctx.message.attachments[0]
+        except KeyError:
```
Comment (Dorukyum) [inline discord/ext/bridge/core.py:390]: The exception is incorrect
```suggestion
        except IndexError:
```

File: discord/ext/bridge/core.py [hunk 16]
```diff
@@ -226,9 +383,14 @@ async def convert(self, ctx, argument):
...
+            raise BadArgument("At least 1 attachment is needed")
```
Comment (Dorukyum) [inline discord/ext/bridge/core.py:391]: Shouldn't a `commands.MissingRequiredArgument` error be raised in this case?

**Review by Dorukyum — CHANGES_REQUESTED** [review]
Hopefully the last review for this pr :sweat_smile:

File: discord/commands/core.py (hunk 14 above)
Comment (Dorukyum) [reply discord/commands/core.py:1070]: A new `TypeVar` bound to `ApplicationContext` could be created for this.

File: discord/ext/bridge/core.py [hunk 17]
```diff
@@ -22,98 +22,165 @@
 FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 DEALINGS IN THE SOFTWARE.
 """
-from typing import Any, List, Union
-import asyncio
 
-import discord.commands.options
-from discord.commands import Option, SlashCommand
-from discord.enums import SlashCommandOptionType
+from __future__ import annotations
```
Comment (BobDotCom) [inline discord/ext/bridge/core.py:26]: Why is this needed?

**Review by Lulalaby — APPROVED** [review]

**Review by Dorukyum — APPROVED** [review]
Good job
